test/object_store/test_backup.py

```diff
@@ -184,3 +186,62 @@ def list_sstables():
     print(f'Check that backup files are still there') # regression test for #20938
     post_objects = set([ o.key for o in get_s3_resource(s3_server).Bucket(s3_server.bucket_name).objects.filter(Prefix=prefix) ])
     assert objects == post_objects
+
+
+@pytest.mark.asyncio
+async def test_backup_and_one_to_one_restore(manager: ManagerClient, s3_server):
+    '''check that 1:1 restoring of a cluster works'''
+
+    cfg = {'enable_user_defined_functions': False,
+           'object_storage_config_file': str(s3_server.config_file),
+           'experimental_features': ['keyspace-storage-options'],
+           'task_ttl_in_seconds': 300
+           }
+    cmd = [ '--logger-log-level', 'sstables_loader=debug:sstable_directory=trace:snapshots=trace:s3=trace:sstable=debug:http=debug' ]
+    servers = await manager.servers_add(3, config=cfg, cmdline=cmd)
+
+    snap_name = unique_name('backup_')
+    logger.info(f'Create and backup keyspace (snapshot name is {snap_name})')
+
+    cql = manager.get_cql()
+    logger.info(f'Create keyspace')
+    ks, cf = create_ks_and_cf(cql, rf=3)
+
+    prefix = f'{cf}/{snap_name}'
+
+    async def do_backup(s):
+        await manager.api.flush_keyspace(s.ip_addr, ks)
+        await manager.api.take_snapshot(s.ip_addr, ks, snap_name)
+        tid = await manager.api.backup(s.ip_addr, ks, cf, snap_name, s3_server.address, s3_server.bucket_name, prefix)
+        await manager.api.wait_task(s.ip_addr, tid)
+
+    await asyncio.gather(*(do_backup(s) for s in servers))
+
+    logger.info(f'Collect sstables lists')
+    sstables = []
+    for s in servers:
+        workdir = await manager.server_get_workdir(s.server_id)
+        cf_dir = os.listdir(f'{workdir}/data/{ks}')[0]
+        tocs = [ f.name for f in os.scandir(f'{workdir}/data/{ks}/{cf_dir}') if f.is_file() and f.name.endswith('TOC.txt') ]
+        sstables.append(tocs)
+    logger.info(f'Sstables: {sstables}')
+
+    logger.info(f'Dropping keyspace')
+    cql.execute(f'DROP KEYSPACE {ks}')
+
+    logger.info(f'Creating ks and cf again')
+    replication_opts = format_tuples({'class': 'NetworkTopologyStrategy', 'replication_factor': '3'})
+    cql.execute((f"CREATE KEYSPACE {ks} WITH REPLICATION = {replication_opts};"))
+    cql.execute(f"CREATE TABLE {ks}.{cf} ( name text primary key, value text );")
+
+    logger.info(f'Restoring')
+    async def do_restore(s, toc_names):
+        logger.info(f'Restore {s.ip_addr} with {toc_names}')
+        tid = await manager.api.restore(s.ip_addr, ks, cf, s3_server.address, s3_server.bucket_name, prefix, toc_names, 'node')
+        await manager.api.wait_task(s.ip_addr, tid)
+
+    await asyncio.gather(*(do_restore(r[0], r[1]) for r in zip(servers, sstables)))
```

Comment:
How does that work with tablets?

Reply:
I don't get the question.
There are three servers, each receives different list of sstables to read mutations from and "node" scope, so each server just downloads its sstables back and that's it

Reply:
I think it works only since rf = |nodes|
Otherwise, if the tablet map of the restored cluster is different than the one of the backed up cluster the application needs to be more careful when restoring the sstables of each node, respectively.  If rf = |racks| it can use the "rack" scope.  And in general, tablet load balancing should be disabled for the duration of restore.

Reply:
> Otherwise, if the tablet map of the restored cluster is different than the one of the backed up cluster the application needs to be more careful when restoring the sstables of each node, respectively

Definitely, but this is what "list of sstables" is to help. Take the corner case example -- load-and-stream with scope=node is called on all nodes giving them full list of sstables from backup. Regardless of how table map looks like, each node will examine all sstables and load all mutations that belong to it according to current state of tablet map.

Reply:
> This is very expensive in communication traffic (time and cost).

It is, but I told that it was a corner case to demonstrate the correctness. In real usage ... (you wrote it already)

> Fast restore should determine the node to load-and-stream from using each sstable token range and original node topology (dc/rack)

There are two options how to get there. First -- caller of restore API filters sstables on its own and feeds smaller subset to each node together with scope=node. This is what I wanted to start from. Second -- sorting in `tablet_sstable_streamer::stream()` checks the scope parameter too and only considers sstables that fit.


Look at it from slightly different angle. Consider we have ... let's call it 1-0.9 restore -- topology is very similar, but still differs, so that there's no 1:1 mapping between sstables in bucket and nodes out there. In this case some sstables will be "just downloaded" and some will be "shared" by nodes either by cherry-picking tokens or by streaming them between nodes. The proposed "scope" approach can handle it as well, so it's pretty universal.

Reply:
If we have multiple nodes re-reading the same sstables that would be costly.

Reply:
The caller supplies list of sstables to each node, so it's possible to make them non-intersecting. In fact, that's the whole point of this PR to facilitate this case. Consider there's 3 nodes, rf=3 and only 1 sstable per node in the backup (3 totally). Current implementation of restore would download them all to a single node (let's call it coordinator) and call load and stream. Coordinator then reads mutations from all three sstables, then spreads the merged result around. With this change each node can be told to download only 1 sstable and "stream" it locally. This is the same amount of bucket->node transferred data, but zero inter-node transfers and no CPU spend by coordinator merging 3 sstables together. Definitely, if caller asks all three nodes to read from all sstables, that's going to be 3x times downloads.

Reply:
>  Current implementation of restore would download them all to a single node (let's call it coordinator)

To my understanding this is inaccurate.
In the example you gave, Scylla Manager splits the 3 stored SStables into 3 buckets (we can assume each bucket is an SStable), and gives each node a bucket to handle. Then, each node download and load and stream the bucket it received from the manger. 
However, there is no attempt whatsoever to try and fit SStables to their corresponding node.
Therefore, it is possible that each node downloads data not belonging to it, and load-and-stream spreads it correctly (in the example with 3 nodes and rf=3 it is not possible, because eventually all nodes will have all tables).